File: tools/tcb_layout/go/parser.go
package tcb_layout

import (
 "encoding/json"
 "errors"
 "os"
 "strings"
)

type FieldDef struct {
 Pos   int    `json:"pos"`
 Len   int    `json:"len"`
 Field string `json:"field"`
 Desc  string `json:"desc"`
}

type LayoutDef struct {
 ID         string     `json:"id"`
 Name       string     `json:"name"`
 TransTypes []string   `json:"trans_types,omitempty"`
 Notes      string     `json:"notes,omitempty"`
 Fields     []FieldDef `json:"fields"`
}

type Spec struct {
 Layouts []LayoutDef `json:"layouts"`
}

type Value struct {
 Raw     string
 Trimmed string
}

type Parsed map[string]Value

func LoadLayouts(path string) ([]LayoutDef, error) {
 b, err := os.ReadFile(path)
 if err != nil {
  return nil, err
 }
 var spec Spec
 if err := json.Unmarshal(b, &spec); err != nil {
  return nil, err
 }
 return spec.Layouts, nil
}

func FindLayout(layouts []LayoutDef, id string) (LayoutDef, bool) {
 for _, l := range layouts {
  if l.ID == id {
   return l, true
  }
 }
 return LayoutDef{}, false
}

func Parse(layout LayoutDef, data []byte) (Parsed, error) {
 if len(data) != 600 {
  return nil, errors.New("payload length must be 600")
 }
 out := make(Parsed, len(layout.Fields))
 for _, f := range layout.Fields {
  start := f.Pos - 1
  end := start + f.Len
  if start < 0 || end > len(data) {
   return nil, errors.New("field slice out of range")
  }
  raw := string(data[start:end])
  out[f.Field] = Value{
   Raw:     raw,
   Trimmed: strings.TrimRight(raw, " "),
  }
 }
 return out, nil
}

File: tools/tcb_layout/ts/parser.ts
import * as fs from "fs";
import * as path from "path";

export interface FieldDef {
  pos: number;
  len: number;
  field: string;
  desc: string;
}

export interface LayoutDef {
  id: string;
  name: string;
  trans_types?: string[];
  notes?: string;
  fields: FieldDef[];
}

export interface Spec {
  layouts: LayoutDef[];
}

export interface Value {
  raw: string;
  trimmed: string;
}

export type Parsed = Record<string, Value>;

export function loadLayouts(jsonPath: string): LayoutDef[] {
  const raw = fs.readFileSync(jsonPath, "utf8");
  const spec = JSON.parse(raw) as Spec;
  return spec.layouts;
}

export function findLayout(layouts: LayoutDef[], id: string): LayoutDef | undefined {
  return layouts.find((l) => l.id === id);
}

export function parse(layout: LayoutDef, data: Uint8Array): Parsed {
  if (data.length !== 600) {
    throw new Error("payload length must be 600");
  }
  const out: Parsed = {};
  for (const f of layout.fields) {
    const start = f.pos - 1;
    const end = start + f.len;
    if (start < 0 || end > data.length) {
      throw new Error("field slice out of range");
    }
    const raw = Buffer.from(data.slice(start, end)).toString("utf8");
    out[f.field] = { raw, trimmed: raw.replace(/\s+$/g, "") };
  }
  return out;
}

export function loadDefaultLayouts(): LayoutDef[] {
  return loadLayouts(path.resolve(__dirname, "../../../docs/tcb_layout_v36.json"));
}
